### PR TITLE
Upgrade Javascript grammar (May 2021)

### PR DESCRIPTION
**Pull request by mjambon**
This upgrades to the latest javascript grammar #3101. Still work in progress.

PR checklist:
- [x] changelog is up to date

### CODE REVIEW

**Comment by mjambon** [comment]
I haven't tested this. If by any chance the tests pass, it should be good to merge. Uses https://github.com/returntocorp/pfff/pull/430

File: CHANGELOG.md [hunk 1]
```diff
@@ -17,6 +17,9 @@ This project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html
 - Evaluate interpolated strings during constant propagation (#3127)
 - Fixed #3084 - Semgrep will report an InvalidRuleSchemaError for dictionaries with duplicate key names.
 - Basic type inference also for implicit variable declarations (Python, Ruby, PHP, and JS)
+- JavaScript parsing: [Support decorators on
```
Comment (brendongo) [inline CHANGELOG.md:20]: Since we cut a release will need to move this changelog entry

Comment (mjambon) [reply CHANGELOG.md:20]: yes, sure. I was hoping to finish last night but didn't make it. The bigger priority for semgrep is the typescript upgrade anyway, which will use some of this work. It should be in the next release unless some complications arise.

File: semgrep-core/src/parsing/tree_sitter/Parse_javascript_tree_sitter.ml [hunk 2]
```diff
@@ -722,20 +793,15 @@ and member_expression (env : env) ((v1, v2, v3) : CST.member_expression) : expr
...
-    | `Choice_choice_get x ->
-        let id = anon_choice_rese_id_9a83200 env x in
-        idexp id
-    | `Choice_obj x -> destructuring_pattern env x
-  in
-  let v2 = token env v2 (* "=" *) in
-  let v3 = expression env v3 in
-  (v1, v2, v3)
+  let parameter = pattern env v1 in
+  let pat = parameter_to_pattern parameter in
```
Comment (aryx) [inline semgrep-core/src/parsing/tree_sitter/Parse_javascript_tree_sitter.ml:797]: we call pattern to get a parameter? and then we convert this parameter (which looks like a pattern ...) to a pattern again?

File: semgrep-core/src/parsing/tree_sitter/Parse_javascript_tree_sitter.ml [hunk 3]
```diff
@@ -1621,6 +1702,46 @@ and anon_choice_pair_5c8e4b4 (env : env) (x : CST.anon_choice_pair_5c8e4b4) :
         }
   | `Semg_dots tok -> todo_semgrep_pattern env tok
 
+(*
+   This is a pattern for destructuring an object property.
```
Comment (aryx) [inline semgrep-core/src/parsing/tree_sitter/Parse_javascript_tree_sitter.ml:1706]: great comment

File: semgrep-core/src/parsing/tree_sitter/Parse_javascript_tree_sitter.ml [hunk 4]
```diff
@@ -1734,23 +1855,18 @@ and declaration (env : env) (x : CST.declaration) : definition list =
       let vars = variable_declaration env x in
       vars_to_defs vars
 
-and formal_parameter (env : env) (x : CST.formal_parameter) : parameter =
+and pattern (env : env) (x : CST.pattern) : parameter =
```
Comment (aryx) [inline semgrep-core/src/parsing/tree_sitter/Parse_javascript_tree_sitter.ml:1858]: ah ok now I get it.

**Review by aryx — APPROVED** [review]
Wow, that's lots of changes.
Thanks for doing that.

File: semgrep-core/src/parsing/tree_sitter/Parse_javascript_tree_sitter.ml (hunk 2 above)
Comment (mjambon) [reply semgrep-core/src/parsing/tree_sitter/Parse_javascript_tree_sitter.ml:797]: We should probably have two different CST-to-AST functions, one that converts to a parameter and another that converts the CST directly to a pattern.

**Pull request merged**